### PR TITLE
Workaround for QEMU bugs causing TLS issues on AArch32

### LINKED ISSUE #52: Revisit ordering of segments in the program header table on AArch32
To fix the issues in #48, @travitch suggests we revisit the segment ordering in https://github.com/GaloisInc/renovate/blob/master/renovate/src/Renovate/BinaryFormat/ELF.hs. Specifically:

> It seems like at the very least `EXIDX` has to come before the first loadable segment, and perhaps even be the first (the documentation I found was unclear)
> It could be the case that we need per-architecture layouts (I hope we can avoid it, but that is the worst case)
> It might be that we can just make sure we leave `EXIDX` as the first segment if it is present
> Ah one important thing to note: `EXIDX` needs to be an early segment, but its *position* in the file (represented by the offset in the output of `readelf`) need not be first.  The ordering of segments is *just* their position in the program headers table

### PR DESCRIPTION
**Pull request by langston-barrett**
Fixes #52 

Based on #48 and #54

### CODE REVIEW

**Comment by langston-barrett** [comment]
Good news: This causes QEMU user-mode to calculate the correct value for `AT_PHDR`. Bad news: The binary still segfaults in TLS-related code, with this backtrace:
```
0x0004bd24 in brk ()
#0  0x0004bd24 in brk ()
#1  0x00029810 in sbrk ()
#2  0x00010920 in __renovate___libc_setup_tls ()
#3  0x000104d4 in __renovate___libc_start_main ()
#4  0x00010288 in _start ()
```
The binary: 
[refurbished-workaround.zip](https://github.com/GaloisInc/renovate/files/4859131/refurbished-workaround.zip)

**Comment by langston-barrett** [comment]
This is almost certainly another discrepancy between Linux and QEMU user-mode. On real ARM hardware running Linux, the binary doesn't crash at that point. In QEMU, the argument (r0) to `brk` is `0x89d08`. In the native case, it's `0xa4d08`. The program header segment is at `0xa3000`, and I assume you'd want the program break to be _after_ the last loadable segment.

**Comment by langston-barrett** [comment]
QEMU bug report no. 2: https://bugs.launchpad.net/qemu/+bug/1886097

tl;dr: Linux calculates the program break based on the virtual address and size of the `LOAD` segment with the highest virtual address, whereas QEMU calculates it based on the virtual address and size of the _writable_ `LOAD` segment with the highest virtual address. Hopefully, we can work around this by just sticking a small writable `LOAD` segment just after `PHDR`.

File: renovate/src/Renovate/BinaryFormat/ELF.hs [hunk 1]
```diff
@@ -465,19 +486,128 @@ doRewrite cfg hdlAlloc loadedBinary symmap strat = do
...
+  let segInfos = mapMaybe makeLoadSegmentInfo phdrs
+  case NEL.nonEmpty segInfos of
+    Nothing -> fail (unlines ("Internal error: No LOAD segments?" : map show phdrs))
+    Just segmentInfos ->
+      case findSpaceForPHDRs segmentInfos projectedOffset requiredSize of
+        Nothing ->
+          -- This is fine in practice if there's no thread-local storage.
+          if null $ filter ((== E.PT_TLS) . E.phdrSegmentType) phdrs
+          then pure defaultAddress
+          else fail $ unlines $
```
Comment (langston-barrett) [inline renovate/src/Renovate/BinaryFormat/ELF.hs:605]: We could also consider making this not fail if the caller explicitly turns on some option saying "I promise not to run this binary in QEMU user-mode".

Comment (langston-barrett) [reply renovate/src/Renovate/BinaryFormat/ELF.hs:605]: @travitch points out that it's hazardous to have separate code paths for binaries we run in QEMU user-mode and those we don't.

**Comment by langston-barrett** [comment]
Status update: This causes the PPC64 tests to fail:
```
$ ~/qemu/ppc64-linux-user/qemu-ppc64 refurbished.exe
refurbished.exe: Invalid argument
```
<details>
  <summary>Details</summary>
I've isolated the problem to `load_elf_image` in `elfload.c` in QEMU, specifically to the `target_mmap` which allocates the virtual memory space. Here are the arguments to that `mmap`:
```
Thread 1 "qemu-ppc64" hit Breakpoint 2, target_mmap (start=start@entry=251047880, len=len@entry=302604344, prot=prot@entry=0, flags=16434, 
    fd=fd@entry=-1, offset=offset@entry=0) at /home/galois/qemu/linux-user/mmap.c:365
365	{
start = 251047880
len = 302604344
prot = 0
flags = 16434
fd = -1
offset = 0
```
and the program headers for the rewritten binary:
```
Program Headers:
  Type           Offset             VirtAddr           PhysAddr
                 FileSiz            MemSiz              Flags  Align
  PHDR           0x000000000012a000 0x000000000f095000 0x000000000f095000
                 0x0000000000000230 0x0000000000000230  R      0x1000
  LOAD           0x000000000012a000 0x000000000f095000 0x000000000f095000
                 0x0000000000000230 0x0000000000000230  R      0x1000
  LOAD           0x0000000000000000 0x0000000010000000 0x0000000010000000
                 0x00000000000b2a11 0x00000000000b2a11  R E    0x10000
  LOAD           0x00000000000b9c80 0x00000000100c9c80 0x00000000100c9c80
                 0x0000000000009b48 0x0000000000009b48  RW     0x10000
  NOTE           0x0000000000000158 0x0000000010000158 0x0000000010000158
                 0x0000000000000044 0x0000000000000044  R      0x4
  TLS            0x00000000000b9c80 0x00000000100c9c80 0x00000000100c9c80
                 0x0000000000000020 0x0000000000000064  R      0x8
  GNU_RELRO      0x00000000000b9c80 0x00000000100c9c80 0x00000000100c9c80
                 0x0000000000006380 0x0000000000006380  R      0x1
  LOAD           0x00000000000d5000 0x000000000f096000 0x000000000f096000
                 0x0000000000052e74 0x0000000000052e74  R E    0x1000
  LOAD           0x0000000000128000 0x0000000020000000 0x0000000020000000
                 0x0000000000001000 0x0000000000001000  RW     0x1000
  LOAD           0x000000000012b038 0x000000000f096000 0x000000000f096000
                 0x0000000000000000 0x0000000000000000   W     0x1000
```
</details>

**Comment by langston-barrett** [comment]
One thing to note is that I introduced a bunch of `fail`s in `doRewrite`, some of which indicate malformed input, others of which indicate either malformed input or an internal bug. When Renovate prints these though, they are prefixed with `User error:`. I could do a separate PR that adds better error handling to the `ElfRewriter` monad if that's desirable.

File: renovate/src/Renovate/BinaryFormat/ELF.hs [hunk 2]
```diff
@@ -465,19 +486,128 @@ doRewrite cfg hdlAlloc loadedBinary symmap strat = do
...
+         , E.elfSegmentVirtAddr = alignedAddr
+         , E.elfSegmentPhysAddr = alignedAddr
+         , E.elfSegmentAlign = 1
+         , E.elfSegmentMemSize = E.ElfRelativeSize 0
+         , E.elfSegmentData = Seq.singleton (E.ElfDataRaw "")
+         }
+
   return analysisResult
 
+filterSegments :: Monad f => (E.ElfSegment w -> Bool) -> E.Elf w -> f [E.ElfSegment w]
```
Comment (langston-barrett) [inline renovate/src/Renovate/BinaryFormat/ELF.hs:527]: This is superfluous (http://fryingpan.dev.galois.com/nix/store/w1bg74najql1gl2hkrn1q0bcs043a9vm-elf-edit-0.35-doc/share/doc/elf-edit-0.35/html/Data-ElfEdit.html#v:elfSegments), removing.

File: refurbish/src/Refurbish/Docker.hs [hunk 3]
```diff
@@ -13,7 +13,7 @@ qemuRunnerName = "refurbish-qemu-runner"
 
 dockerFile :: String
 dockerFile =
-  unlines [ "FROM ubuntu:18.04"
+  unlines [ "FROM ubuntu:20.04"
```
Comment (langston-barrett) [inline refurbish/src/Refurbish/Docker.hs:16]: Just FYI, this was necessary for some of the tests to pass (we needed a newer version of QEMU).

File: renovate/src/Renovate/BinaryFormat/ELF.hs (hunk 1 above)
Comment (langston-barrett) [reply renovate/src/Renovate/BinaryFormat/ELF.hs:611]: I think we're leaving this as-is for now, "most" binaries seem to pass at this point.

**Comment by langston-barrett** [comment]
Okay, this is cleaned up enough to be ready for review.

File: refurbish/src/Refurbish/Docker.hs (hunk 3 above)
Comment (travitch) [reply refurbish/src/Refurbish/Docker.hs:16]: Oh boy, that sounds exciting.  Do you happen to know the required version of qemu? It would be helpful to have that recorded in maybe the README

Comment (langston-barrett) [reply refurbish/src/Refurbish/Docker.hs:16]: I just tried with the version I compiled from `master`, then switched this version and it noticed that it works with this package too... The version in 20.04 is 4.2-3ubuntu6.1, and in 18.04 it's 2.11+dfsg-1ubuntu7.26. I could try to bisect, or we could leave it at that?

Comment (travitch) [reply refurbish/src/Refurbish/Docker.hs:16]: Known to work on 4.2 is reasonable, no need to bisect

Comment (langston-barrett) [reply refurbish/src/Refurbish/Docker.hs:16]: I'll add a line in the README.

File: renovate/src/Renovate/BinaryFormat/ELF/Internal.hs [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+                            alignValue (pVAddr loSeg + pMemSz loSeg + 1) pgAlign
+                    , phdrSize < pVAddr hiSeg - addr
+                    ]
+                  , [ addr
+                    | (loSeg, hiSeg) <- zipped
+                    , let addr =
+                            alignValueDown (pVAddr hiSeg - phdrSize - 1) pgAlign
+                    , addr > pVAddr loSeg + pMemSz loSeg
+                    ]
+                  ]
```
Comment (travitch) [inline renovate/src/Renovate/BinaryFormat/ELF/Internal.hs:110]: This bit looks fairly subtle - can we factor out some bits with more descriptive names, or just give a textual description?

File: renovate/src/Renovate/BinaryFormat/ELF/Internal.hs [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+      rawCandidates = Maybe.maybeToList beforeFirst ++ between ++ afterLast
+      validCandidates = filter (>= phdrOffset) rawCandidates
+  in
+     if length validCandidates == 0
```
Comment (travitch) [inline renovate/src/Renovate/BinaryFormat/ELF/Internal.hs:115]: `null`?

File: renovate/src/Renovate/BinaryFormat/ELF/Internal.hs [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+     then Nothing
+     else
+       let best = L.minimumBy (O.comparing (\addr -> abs (addr - phdrOffset))) validCandidates
+       in if let m = minimum (fmap (\segInfo -> abs (pVAddr segInfo - pOffset segInfo)) segInfos)
```
Comment (travitch) [inline renovate/src/Renovate/BinaryFormat/ELF/Internal.hs:119]: Is this essentially the calculation that QEMU usermode is doing?

File: renovate/src/Renovate/BinaryFormat/ELF.hs [hunk 7]
```diff
@@ -450,12 +452,39 @@ doRewrite cfg hdlAlloc loadedBinary symmap strat = do
   -- Now overwrite the original code (in the .text segment) with the
   -- content computed by our transformation.
   modifyCurrentELF (overwriteTextSection overwrittenBytes)
-  let newSegment = phdrSegment phdrSegmentAddress
+
+  -- Add a PHDR segment which will hold the modified program header table.
+  --
+  -- We usually want this to be the first segment (have a segment index of 0)
+  -- (see comment on "phdrSegment"), except that on ARM, if an EXIDX segment is
+  -- present, that must come first.
```
[the remark this answers is not in the document]
Comment (langston-barrett) [reply renovate/src/Renovate/BinaryFormat/ELF.hs:460]: It does also have to be first. The "minimal difference constraint" comes from QEMU, the "PHDRs are the first LOAD segment" constraint comes from the kernel (recall that there's that loop where `load_addr` is set, and it's only set according to data from the first `LOAD` segment, and it flows into `AT_PHDR).

File: renovate/src/Renovate/BinaryFormat/ELF.hs [hunk 8]
```diff
@@ -450,12 +452,39 @@ doRewrite cfg hdlAlloc loadedBinary symmap strat = do
...
+
+  let isEXIDX seg = E.elfSegmentType seg == E.PT_ARM_EXIDX
+  exidxs <- withCurrentELF (pure . filter isEXIDX . E.elfSegments)
+  when (length exidxs > 1) $
+    fail $ unwords $
```
[the remark this answers is not in the document]
Comment (langston-barrett) [reply renovate/src/Renovate/BinaryFormat/ELF.hs:465]: Sure!

File: renovate/src/Renovate/BinaryFormat/ELF.hs [hunk 9]
```diff
@@ -450,12 +452,39 @@ doRewrite cfg hdlAlloc loadedBinary symmap strat = do
...
+      Nothing -> pure 0
+      Just idx | idx == 0 -> pure 1
+      Just idx ->
+        fail $ unwords $
+          [ "Malformed input: "
+          , "Expected EXIDX segment to have index 0, but it had index"
+          , show idx
+          ]
+
+  phdrSegmentAddress <- withCurrentELF (choosePHDRSegmentAddress Proxy)
```
[the remark this answers is not in the document]
Comment (langston-barrett) [reply renovate/src/Renovate/BinaryFormat/ELF.hs:481]: Good catch. Looks like that was left over from a previous version.

File: renovate/src/Renovate/BinaryFormat/ELF.hs [hunk 10]
```diff
@@ -465,19 +494,126 @@ doRewrite cfg hdlAlloc loadedBinary symmap strat = do
...
+              [ "Internal error:"
+              , "Wrong number of PT_PHDR segments:"
+              , show phdrs'
+              ]
+
+  nextIdx <- withCurrentELF (pure . (+1) . nextSegmentIndex)
+  modifyCurrentELF $ appendSegment $
+    let alignedAddr =
+          alignValue
+            (E.elfSegmentVirtAddr newSegment + phdrActualSize)
```
[the remark this answers is not in the document]
Comment (langston-barrett) [reply renovate/src/Renovate/BinaryFormat/ELF.hs:523]: The comment above this section says:
```
  -- the highest virtual address. Since our PHDR segment might have a high
  -- virtual address, we have to add a writable load segment after it, so our
  -- programs don't segfault inside QEMU.
```
But I will also add a line comment here specifically.

File: renovate/src/Renovate/BinaryFormat/ELF.hs [hunk 11]
```diff
@@ -465,19 +494,126 @@ doRewrite cfg hdlAlloc loadedBinary symmap strat = do
...
+        case filter ((== E.PT_PHDR) . E.phdrSegmentType) (E.allPhdrs (E.elfLayout elf)) of
+          [seg] -> pure seg
+          phdrs' ->
+            fail $ unwords $
+              [ "Internal error:"
+              , "Wrong number of PT_PHDR segments:"
+              , show phdrs'
+              ]
+
+  nextIdx <- withCurrentELF (pure . (+1) . nextSegmentIndex)
```
[the remark this answers is not in the document]
Comment (langston-barrett) [reply renovate/src/Renovate/BinaryFormat/ELF.hs:519]: Yes, great point. Just laziness!

File: renovate/src/Renovate/BinaryFormat/ELF.hs [hunk 12]
```diff
@@ -465,19 +494,126 @@ doRewrite cfg hdlAlloc loadedBinary symmap strat = do
...
+  -- Linux calculates the program break based on the virtual address and size of
+  -- the LOAD segment with the highest virtual address, whereas QEMU calculates
+  -- it based on the virtual address and size of the *writable* LOAD segment with
+  -- the highest virtual address. Since our PHDR segment might have a high
+  -- virtual address, we have to add a writable load segment after it, so our
+  -- programs don't segfault inside QEMU.
+  --
+  -- https://bugs.launchpad.net/qemu/+bug/1886097
+
+  phdrActualSize <-
```
[the remark this answers is not in the document]
Comment (langston-barrett) [reply renovate/src/Renovate/BinaryFormat/ELF.hs:507]: Yes, good idea!

File: renovate/src/Renovate/BinaryFormat/ELF.hs [hunk 13]
```diff
@@ -315,13 +312,18 @@ sectionAddressRange sec = (textSectionStartAddr, textSectionEndAddr)
...
+--
 -- The strategy we choose is to:
 -- 1) Wipe out the original PHDRs table (along with all of the other dynamically-sized data)
 -- 2) Append the new text segment (with the .extratext section) after all of the existing segments
 -- 3) Append new copies of all of the dynamically-sized sections (including a new symbol table)
 -- 4) Append a fresh PHDRs table at the very end *but* with segment index 0 (with the other segment
---    indexes suitably modified), which makes it the first loadable segment.  We assign a very high
---    address to the PHDRs segment so that it will always be greater than the file offset.
+--    indexes suitably modified), which makes it the first loadable segment.
+-- 5) Append an empty, writable, loadable segment at an address just after the new PHDRs.
```
[the remark this answers is not in the document]
Comment (langston-barrett) [reply renovate/src/Renovate/BinaryFormat/ELF.hs:326]: Will do.

File: renovate/src/Renovate/BinaryFormat/ELF.hs [hunk 14]
```diff
@@ -519,13 +655,17 @@ appendSegment seg e = do
...
 --
--- We choose 0 as the segment index, so ensure that we've already made space in
--- the segment index space for it (by e.g., incrementing all of the other
--- segment indexes).
+-- Requires that we've already made space in the segment index space (by e.g.,
+-- incrementing all of the other segment indexes past @idx@).
+--
+-- The spec says that if there's a PHDR segment, it must be before any loadable
+-- segment in the table, so loadable segments should all have indices greater
+-- than @idx@.
```
[the remark this answers is not in the document]
Comment (langston-barrett) [reply renovate/src/Renovate/BinaryFormat/ELF.hs:663]: I'm not sure if it's because of a specification, but it's definitely necessary (see above comment)

File: renovate/src/Renovate/BinaryFormat/ELF.hs [hunk 15]
```diff
@@ -544,14 +684,22 @@ phdrSegment addr =
...
 
-incrementSegmentNumber :: (Monad m) => E.SegmentIndex -> E.ElfSegment w -> m (E.ElfSegment w)
-incrementSegmentNumber n seg = return seg { E.elfSegmentIndex = E.elfSegmentIndex seg + n }
+incrementSegmentNumber ::
+  (Monad m) =>
+  (E.ElfSegment w -> Bool) ->
+  E.SegmentIndex ->
+  E.ElfSegment w ->
+  m (E.ElfSegment w)
+incrementSegmentNumber predicate n seg =
```
[the remark this answers is not in the document]
Comment (langston-barrett) [reply renovate/src/Renovate/BinaryFormat/ELF.hs:698]: Sure!

File: renovate/src/Renovate/BinaryFormat/ELF/Internal.hs (hunk 5 above)
Comment (langston-barrett) [reply renovate/src/Renovate/BinaryFormat/ELF/Internal.hs:115]: :+1:

File: renovate/src/Renovate/BinaryFormat/ELF/Internal.hs (hunk 6 above)
Comment (langston-barrett) [reply renovate/src/Renovate/BinaryFormat/ELF/Internal.hs:119]: Yep, exactly. This is a last-ditch check to ensure that this function doesn't return a `Just addr` where `addr` would be unacceptable to QEMU.

File: renovate/src/Renovate/BinaryFormat/ELF.hs [hunk 16]
```diff
@@ -450,12 +452,39 @@ doRewrite cfg hdlAlloc loadedBinary symmap strat = do
...
+
+  let isEXIDX seg = E.elfSegmentType seg == E.PT_ARM_EXIDX
+  exidxs <- withCurrentELF (pure . filter isEXIDX . E.elfSegments)
+  when (length exidxs > 1) $
+    fail $ unwords $
+      [ "Malformed input:"
+      , "Found several ARM_EXIDX segments, at the following indices:"
+      , show (map E.elfSegmentIndex exidxs)
+      ]
+  newSegmentIdx <-
```
[the remark this answers is not in the document]
Comment (langston-barrett) [reply renovate/src/Renovate/BinaryFormat/ELF.hs:470]: Yep, good choice!

File: renovate/src/Renovate/BinaryFormat/ELF.hs [hunk 17]
```diff
@@ -465,19 +494,126 @@ doRewrite cfg hdlAlloc loadedBinary symmap strat = do
...
+              , "Wrong number of PT_PHDR segments:"
+              , show phdrs'
+              ]
+
+  nextIdx <- withCurrentELF (pure . (+1) . nextSegmentIndex)
+  modifyCurrentELF $ appendSegment $
+    let alignedAddr =
+          alignValue
+            (E.elfSegmentVirtAddr newSegment + phdrActualSize)
+            (fromIntegral pageAlignment)
```
[the remark this answers is not in the document]
Comment (langston-barrett) [reply renovate/src/Renovate/BinaryFormat/ELF.hs:524]: ... I don't know, but it works? :grimacing:

File: renovate/src/Renovate/BinaryFormat/ELF.hs (hunk 10 above)
[the remark this answers is not in the document]
Comment (langston-barrett) [reply renovate/src/Renovate/BinaryFormat/ELF.hs:509]: https://github.com/GaloisInc/renovate/pull/55/commits/7d80396db4fe23e9ae4a6592450f7963dcb253e3

File: renovate/src/Renovate/BinaryFormat/ELF.hs (hunk 13 above)
[the remark this answers is not in the document]
Comment (langston-barrett) [reply renovate/src/Renovate/BinaryFormat/ELF.hs]: https://github.com/ge-high-assurance/RACK/pull/77#event-3567012581

File: renovate/src/Renovate/BinaryFormat/ELF.hs (hunk 9 above)
[the remark this answers is not in the document]
Comment (langston-barrett) [reply renovate/src/Renovate/BinaryFormat/ELF.hs]: https://github.com/GaloisInc/renovate/pull/55/commits/7d80396db4fe23e9ae4a6592450f7963dcb253e3

File: renovate/src/Renovate/BinaryFormat/ELF.hs (hunk 8 above)
[the remark this answers is not in the document]
Comment (langston-barrett) [reply renovate/src/Renovate/BinaryFormat/ELF.hs]: https://github.com/GaloisInc/renovate/pull/55/commits/7d80396db4fe23e9ae4a6592450f7963dcb253e3

File: renovate/src/Renovate/BinaryFormat/ELF.hs (hunk 15 above)
[the remark this answers is not in the document]
Comment (langston-barrett) [reply renovate/src/Renovate/BinaryFormat/ELF.hs]: https://github.com/GaloisInc/renovate/pull/55/commits/7d80396db4fe23e9ae4a6592450f7963dcb253e3

File: renovate/src/Renovate/BinaryFormat/ELF/Internal.hs (hunk 5 above)
Comment (langston-barrett) [reply renovate/src/Renovate/BinaryFormat/ELF/Internal.hs]: https://github.com/GaloisInc/renovate/pull/55/commits/7d80396db4fe23e9ae4a6592450f7963dcb253e3

File: renovate/src/Renovate/BinaryFormat/ELF/Internal.hs (hunk 4 above)
Comment (langston-barrett) [reply renovate/src/Renovate/BinaryFormat/ELF/Internal.hs:122]: I added some detailed commentary

**Comment by langston-barrett** [comment]
@travitch What's your preference on how we merge this? I don't think cleaning up the git history is super feasible, and there's a lot of content... Should I squash? Rebase and try to get halfway coherent commits?

**Comment by travitch** [comment]
Good question.  I'm happy with a squash commit or, alternatively, you could do a local rebase (possibly into one commit?), force push to the branch, and then do a normal merge?
